fix: manually write value to /proc/sys/net/netfilter/nf_conntrack_max

toabctl opened the pull request:
That's required because the nf_conntrack module gets loaded after systemd-sysctl.service gets executed so the value written to /etc/sysctl.d/50-aws-gadget.conf does not get applied after first boot (LP:#2125739).

toabctl committed:
fix: manually write value to /proc/sys/net/netfilter/nf_conntrack_max

That's required because the nf_conntrack module gets loaded after
systemd-sysctl.service gets executed so the value written to
/etc/sysctl.d/50-aws-gadget.conf does not get applied after first
boot (LP:#2125739).

toabctl commented:
relevant store-request: https://forum.snapcraft.io/t/more-privileged-interfaces-for-strict-aws-gadget-snap/48761

alfonsosanchezbeato commented on snap/hooks/install-device:
```diff
@@ -9,3 +9,6 @@ net.ipv4.neigh.default.gc_thresh2=15360
 EOF
+
+# NOTE: keep this in sync with the value written to 50-aws-gadget.conf
+echo 1048576 > /proc/sys/net/netfilter/nf_conntrack_max
```
But this would be temporary for the install boot, then you reboot into normal run mode. Note also that when eventually we start to use preseeding, you do not want an install-device hook as it prevents installation from the initramfs. This should be in the install and post-refresh hooks (the latter in case the setting changes for a new gadget). After first boot in run mode, sysctl configuration should be enough.

toabctl replied:
done (only `install` hook for now)

toabctl commented:
@alfonsosanchezbeato could you have another look please?

alfonsosanchezbeato commented on snapcraft.yaml:
```diff
@@ -28,6 +28,11 @@ plugs:
     write:
       - /etc/sysctl.d/50-aws-gadget.conf
+  # workaround because the nf_conntrack module gets loaded after systemd-sysctl.service runs (LP:#2125739)
+  proc-sys-net-netfilter-nf_conntrack_max:
```
Underscores are not allowed in dictionary keys by snapd, this should be `proc-sys-net-netfilter-nf-conntrack-max` to avoid being ignored by snapd.

toabctl replied:
fixed

alfonsosanchezbeato commented on snapcraft.yaml:
```diff
@@ -38,6 +43,9 @@ hooks:
   install-device:
     plugs:
       - install-mode-etc-sysctl-d-50-aws-gadget-conf
+  install-:
```
There is an extra `-` at the end

toabctl replied:
fixed

alfonsosanchezbeato commented on snapcraft.yaml:
```diff
@@ -38,6 +43,9 @@ hooks:
       - install-mode-etc-sysctl-d-50-aws-gadget-conf
+  install-:
+    plugs:
+      - proc-sys-net-netfilter-nf_conntrack_max
```
Same: re plug name

toabctl replied:
fixed

alfonsosanchezbeato commented on snap/hooks/install:
I wonder if maybe is worth it to both write `50-aws-gadget.conf` and write the settings  to /proc/sys/  from the install hook. Then you could remove the install-device hook - the only benefit that you are getting from it is that the kernel options are set early on first "run" boot, which might be not such a greate benefit. That would simplify the gadget a bit.

Also, note that post-refresh is only relevant on a snap refresh, so what you are doing there atm would overwrite an existing file. It might be needed if a new snap wants to set different kernel options/values.

toabctl replied:
but then I would need to adjust the store request to allow to write to 
* /proc/sys/net/ipv4/neigh/default/gc_thresh2
* /proc/sys/net/ipv4/neigh/default/gc_thresh3

. That's also an option but if future changes are required, I would create another store request for each change. 
But you already mentioned that the `install-device` hook will have to go away if we want to use preseeded images so that makes sense. thx!

toabctl replied:
updated. and yes, the `post-refresh` is there if we have to fix bugs with the current snap.

alfonsosanchezbeato reviewed (commented):
Thanks, some comments

toabctl committed:
feat: restructure hooks to avoid install-device hook

The install-device hook doesn't get executed in preseeded images. The
future plan is to use preseeding to improve first boot speed.

toabctl committed:
fix: switch to network-control interface

That's the recommentation from the store-request[0] given that the
network-control interface will allow to write to
@{PROC}/sys/net/ipv{4,6}/** .

[0]
https://forum.snapcraft.io/t/more-privileged-interfaces-for-strict-aws-gadget-snap/48761/7

alfonsosanchezbeato reviewed (approved):
LGTM
